photon-ml/src/main/scala/com/linkedin/photon/ml/optimization/DistributedOptimizationProblem.scala

```diff
@@ -0,0 +1,198 @@
+/*
+ * Copyright 2016 LinkedIn Corp. All rights reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package com.linkedin.photon.ml.optimization
+
+import breeze.linalg.Vector
+import org.apache.spark.broadcast.Broadcast
+import org.apache.spark.rdd.RDD
+
+import com.linkedin.photon.ml.constants.{MathConst, StorageLevel}
+import com.linkedin.photon.ml.data.LabeledPoint
+import com.linkedin.photon.ml.function.{DistributedObjectiveFunction, L2Regularization, TwiceDiffFunction}
+import com.linkedin.photon.ml.model.Coefficients
+import com.linkedin.photon.ml.normalization.NormalizationContext
+import com.linkedin.photon.ml.sampler.DownSampler
+import com.linkedin.photon.ml.supervised.model.{GeneralizedLinearModel, ModelTracker}
+
+/**
+ * An optimization problem solved by multiple tasks on one or more executors. Used for solving the global optimization
+ * problems of a fixed effect model.
+ *
+ * @param optimizer The underlying optimizer which iteratively solves the convex problem
+ * @param objectiveFunction The objective function to optimize
+ * @param samplerOption (Optional) A sampler to use for down-sampling the training data prior to optimization
+ * @param glmConstructor The function to use for producing GLMs from trained coefficients
+ * @param regularizationContext The regularization context
+ * @param isComputingVariances Should coefficient variances be computed in addition to the means?
+ */
+protected[ml] class DistributedOptimizationProblem[Function <: DistributedObjectiveFunction] protected[optimization] (
```

Comment:
Split the optimization problems into two variants: IndividualOptimizationProblem and DistributedOptimizationProblem.

IndividualOptimizationProblems are used to solve the individual random effect optimization problems local to one task.

DistributedOptimizationProblems are used to solve fixed effect optimization problems across several tasks on one or more executors. They can also downsample input data prior to each run.

This split helps to enforce Photon/GAME as a pipeline of transformers. It also makes it easier to quickly determine which classes should be serializable and which should not.